### ISSUE TITLE
huge amount of comments requests being made to GitHub API

### ISSUE THREAD

**Issue by boegel** [janinko/ghprb#333]
We're seeing issues with huge amounts of comment fetch requests being made to the GitHub API (causing our Jenkins server to continuously hit the rate limit of 5000 requests per hour).

From the Jenkins log, triggered by a *single* comment being made in a pull request:

```
Sep 23, 2015 2:16:06 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:06 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:07 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:07 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:07 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:07 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:07 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:07 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:07 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:07 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:07 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:07 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:08 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:08 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:08 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:08 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:08 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:08 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:08 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:08 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:08 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:08 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:09 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:09 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:09 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:09 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Checking issue comment 'org.kohsuke.github.GHIssueComment@18c7ddd8' for repo hpcugent/easybuild-easyconfigs
Sep 23, 2015 2:16:09 PM INFO org.jenkinsci.plugins.ghprb.GhprbPullRequest updatePR
Pull request #1,001 was updated on hpcugent/easybuild-easyconfigs at 9/23/15 2:16 PM by User:boegel
Sep 23, 2015 2:16:10 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:10 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:10 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:10 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:10 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:11 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:11 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
Sep 23, 2015 2:16:11 PM INFO org.jenkinsci.plugins.ghprb.GhprbWebHook handleWebHook
Got payload event: issue_comment
```

Is there any way to trim this down further, and figure out what's going on? Are there detailed logs available for ghprb somewhere?

**Comment by DavidTanner** [comment]
There does need to be a better way to identify what the webHook is, but it shouldn't be calling to GitHub with each of those checks.

**Comment by boegel** [comment]
OK, this may be just the logging that's a bit too verbose, sure.

But there's definitely a problem.

GitHub support tells us that ghpbr has been requesting the comments for https://github.com/hpcugent/easybuild-easyconfigs/pull/1943 about 200 times in the last couple of hours, but that pull requests hasn't been updated in 15 days.

Does ghpbr cache things somewhere in a file that we can clean up, or something like that?

**Comment by DavidTanner** [comment]
can you get the URL that is being queried?

**Comment by boegel** [comment]
@DavidTanner: GitHub support told us this is the target (for example, there are other PRs being queried too like this):

```
/repos/hpcugent/easybuild-easyconfigs/pulls/1943/comments
```

**Comment by DavidTanner** [comment]
In the logs you posted the issue_comment is a webhook sent FROM GitHub TO ghrpb for PR #1001.  It doesn't make any calls TO GitHub to handle the hook, but it does check every job to see if it matches the repo in the hook.

**Comment by boegel** [comment]
@DavidTanner: sure, I understand that, and it may not be related to the large amount of API requests being made, but it seems awfully related from a high level...

That doesn't take away that, for some reason, ghpbr is querying the comments in the last couple of hours for a PR that hasn't been updated in over 2 weeks...

**Comment by boegel** [comment]
@DavidTanner: how destructive is removing `org.jenkinsci.plugins.ghprb.GhprbTrigger.xml` (and then restarting Jenkins), where I can see 1943 being listed in?

**Comment by DavidTanner** [comment]
https://github.com/jenkinsci/ghprb-plugin/blob/ghprb-1.28.3/src/main/java/org/jenkinsci/plugins/ghprb/GhprbPullRequest.java#L340 is the only line that checks the comments for a pull request.  From the logs for webhook handling it looks like you have a lot of jobs, and if each of these are also set up with a cron trigger then they are each going to check on the pull requests they are responsible for every time the cron runs.

**Comment by DavidTanner** [comment]
If you remove the xml file, then the plugin will start with every PR as a new one, and the same problem will happen.  If you are using webhooks, then you should set the cron to something that will never run.  I have used ```* * 30 2 * *```, which would be 2/30.

**Comment by boegel** [comment]
We only have one job in place that uses ghpbr (for this easybuild-easyconfigs repo, we do have jobs for other repos), and it's not using a cron setup at all, it's configured such that only events on GitHub trigger it (a new pull request, or an issue comments mentioning something like "test this please").

**Comment by boegel** [comment]
@DavidTanner: maybe some default setting for ghprb changed recently, causing it to requery open pull requests? We do have a lot of them (~290)...

**Comment by DavidTanner** [comment]
Check the cron setting, just out of curiosity.

**Comment by boegel** [comment]
I just noticed this in the job log that uses ghprb, maybe that explains things for you somehow?
It certainly points at some kind of issue...

```
GitHub pull request #931 of commit e35ea3f24d458b92c8992bf43a20f06b8fc0564a, no merge conflicts.
Unable to query GitHub for status of PullRequestjava.lang.NullPointerException
	at org.jenkinsci.plugins.ghprb.GhprbBuilds.onStarted(GhprbBuilds.java:77)
	at org.jenkinsci.plugins.ghprb.GhprbBuildListener.onStarted(GhprbBuildListener.java:19)
	at org.jenkinsci.plugins.ghprb.GhprbBuildListener.onStarted(GhprbBuildListener.java:12)
	at hudson.model.listeners.RunListener.fireStarted(RunListener.java:213)
	at hudson.model.Run.execute(Run.java:1734)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:381)
Setting status of e35ea3f24d458b92c8992bf43a20f06b8fc0564a to PENDING with url 
```

**Comment by boegel** [comment]
@DavidTanner: this is the top part of the ghpbr XML config file:

```
<?xml version='1.0' encoding='UTF-8'?>
<org.jenkinsci.plugins.ghprb.GhprbTrigger_-DescriptorImpl plugin="ghprb@1.27">
  <configVersion>1</configVersion>
  <whitelistPhrase>.*add\W+to\W+whitelist.*</whitelistPhrase>
  <okToTestPhrase>.*[oO][kK]\W+to\W+test.*</okToTestPhrase>
  <retestPhrase>.*test\W+this.*</retestPhrase>
  <skipBuildPhrase>.*\[skip\W+ci\].*</skipBuildPhrase>
  <cron>H/5 * * * *</cron>
  <useComments>false</useComments>
  <useDetailedComments>false</useDetailedComments>
  <unstableAs>FAILURE</unstableAs>
  <autoCloseFailedPullRequests>false</autoCloseFailedPullRequests>
  <displayBuildErrorsOnDownstreamBuilds>false</displayBuildErrorsOnDownstreamBuilds>
  <githubAuth>
```

I guess the `cron` part confirms your suspicion?

**Comment by jlebon** [comment]
@boegel You can change the setting for it in the job configuration if you click on the "Advanced" button. There is also a default setting for it in the Jenkins settings (also under "Advanced").

**Comment by DavidTanner** [comment]
Which version of ghprb are you currently using?  I want to find that line that is throwing the null pointer.

**Comment by boegel** [comment]
@DavidTanner: see `plugin="ghprb@1.27"` above?

**Comment by DavidTanner** [comment]
Sorry, missed it.

**Comment by boegel** [comment]
@jlebon: we've been using ghprb for a *long* time, so it's beyond me why this suddenly pops up now...

Maybe @JensTimmerman knows, he updated ghprb recently after #319 was fixed ...

**Comment by DavidTanner** [comment]
I am not sure why the PR is null at that line.  Can you post the entire build log?
https://github.com/jenkinsci/ghprb-plugin/blob/ghprb-1.27/src/main/java/org/jenkinsci/plugins/ghprb/GhprbBuilds.java#L77

**Comment by boegel** [comment]
@DavidTanner: the bit I showed is the top of the build log...

**Comment by JensTimmerman** [comment]
Negative, I updated to 1.27 and then reported back in #319, we have the latest release, not a dev version.

**Comment by JensTimmerman** [comment]
our crontab line said `* * 30 2 *` Which should never run, but jenkins was asking if we really meant every minute, so changed to `H * 30 2 *` to get rid of that warning, But this should not make a difference?

**Comment by boegel** [comment]
Hmm, hold on, not out of the woods yet, I think...

The `cron` entry in `org.jenkinsci.plugins.ghprb.GhprbTrigger.xml` is just the *default*, right?

If I check our `easybuild-easyconfigs` pull request builder job via the Jenkins interface, I see that 'Build periodically' is not enabled, and if I do select it (without saving), the input box is empty and a nice "No schedules so will never run" warning is shown.

So, the job is configured without a cron, right?

However, the `config.xml` for the corresponding job shows this:

```
jobs/easybuild-easyconfigs-pr-builder/config.xml:      <cron>H/5 * * * *</cron>
```

So, is the cron enabled, or not? Is there another setting in the `config.xml` that corresponds to having the cron enabled/disabled? I'm very confused now...

**Comment by boegel** [comment]
@JensTimmerman: it said `* * 30 2 *` after I changed it 30m ago...

**Comment by boegel** [comment]
So, apparently, when I changed the 'global' cron setting, the `H/5 * * * *` value was just copied down to each of the jobs, where I had to modify it as well.

Is this the intended behaviour?

**Comment by DavidTanner** [comment]
the default for the config page is the global.cron value, but that only gets set if there is nothing in the cron variable locally.  Each of the jobs keep and set their own value.

**Comment by DavidTanner** [comment]
If you saved the trigger with a blank cron then the default would show up.

**Comment by JensTimmerman** [comment]
DavidTanner, trying to save `* * 30 2 * *`  gives an exception, (since there's an extra `*`  ?) 
Could this also result in the default showing up everywhere?

**Comment by boegel** [comment]
The `cron` settings get changed, but they're not used in our setup, as far as I can tell.

I'm not seeing anything in the Jenkins log w.r.t. ghprb being triggered regularly via cron at all, I guess it should be mentioned? I only see ghprb being triggered via the web hook.

I think something else is going on, it seems that the ghprb version we're running now is revisiting all jobs in the build history (without being triggered by a cron)...
This explains why GitHub support mentioned that we were fetching comments for an old PR that wasn't updated in 15 days; we were keeping all job results for 15 days (I now changed this to just keep the last 10 jobs, to see if this helps with avoiding hitting the rate limit).

@DavidTanner: is there a way to get more detailed logging from ghprb? There's now very little information about what it's doing (for example, which requests it's making to the GitHub API, and why).

**Comment by boegel** [comment]
An update from our side: as it turns out, the problems we were seeing with ghprb were a result rather than a cause.

It wasn't ghprb that was hammering the GitHub API, but our Jenkins instance (which uses the same GitHub account as our Jenkins instance does, via tokens).

Sorry for all the confusion...
